types/module/module.go

```diff
@@ -17,89 +18,127 @@ process. This separation is necessary, however we still want to allow for a
 high level pattern for modules to follow - for instance, such that we don't
 have to manually register all of the codecs for all the modules. This basic
 procedure as well as other basic patterns are handled through the use of
-ModuleBasicManager.
+BasicManager.
+
+Lastly the interface for genesis functionality (AppModuleGenesis) has been
+separated out from full module functionality (AppModule) so that modules which
+are only used for genesis can take advantage of the Module patterns without
+needlessly defining many placeholder functions
 */
-package types
+package module
 
 import (
 	"encoding/json"
 
-	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/gorilla/mux"
 	"github.com/spf13/cobra"
+
 	abci "github.com/tendermint/tendermint/abci/types"
-)
 
-// ModuleClient helps modules provide a standard interface for exporting client functionality
-type ModuleClient interface {
-	GetQueryCmd() *cobra.Command
-	GetTxCmd() *cobra.Command
-}
+	"github.com/cosmos/cosmos-sdk/client/context"
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
 
 //__________________________________________________________________________________________
 // AppModuleBasic is the standard form for basic non-dependant elements of an application module.
 type AppModuleBasic interface {
 	Name() string
 	RegisterCodec(*codec.Codec)
+
+	// genesis
 	DefaultGenesis() json.RawMessage
 	ValidateGenesis(json.RawMessage) error
+
+	// client functionality
+	RegisterRESTRoutes(context.CLIContext, *mux.Router, *codec.Codec)
```

Comment:
Out of scope of this PR, but we can and should stop passing a `CLIContext` AND a `*coded.Codec`. The context should already have the codec set. I'll make an issue of this and address it once this PR is merged.

REF: https://github.com/cosmos/cosmos-sdk/issues/4479